pyro/distributions/transforms/block_naf.py

```diff
@@ -0,0 +1,232 @@
+            raise ValueError('Invalid activation function "{}"'.format(activation))
+        self.f = name_to_mixin[activation].f
+        self.log_df_dx = name_to_mixin[activation].log_df_dx
+
+        # Initialize modules for each layer in flow
+        self.residual = residual
+        self.input_dim = input_dim
+        self.layers = nn.ModuleList([MaskedBlockLinear(input_dim, input_dim * hidden_factors[0], input_dim)])
+        for idx in range(1, len(hidden_factors)):
+            self.layers.append(MaskedBlockLinear(
+                input_dim * hidden_factors[idx - 1], input_dim * hidden_factors[idx], input_dim))
+        self.layers.append(MaskedBlockLinear(input_dim * hidden_factors[-1], input_dim, input_dim))
+        self._cached_logDetJ = None
+
+        if residual == 'gated':
+            self.gate = torch.nn.Parameter(torch.nn.init.normal_(torch.Tensor(1)))
+
+    def _call(self, x):
+        """
+        :param x: the input into the bijection
+        :type x: torch.Tensor
+
+        Invokes the bijection x=>y; in the prototypical context of a TransformedDistribution `x` is a
+        sample from the base distribution (or the output of a previous flow)
+        """
+        y = x
+        for idx in range(len(self.layers)):
+            pre_activation, dy_dx = self.layers[idx](y.unsqueeze(-1))
+            J_act = self.log_df_dx(pre_activation).view(*(list(x.size()) + [-1, 1]))
+
+            if idx == 0:
+                logDetJ = dy_dx + J_act
+                y = self.f(pre_activation)
+            elif idx < len(self.layers) - 1:
+                logDetJ = log_matrix_product(dy_dx, logDetJ) + J_act
+                y = self.f(pre_activation)
+            else:
+                logDetJ = log_matrix_product(dy_dx, logDetJ)
+                y = pre_activation
+
+        self._cached_logDetJ = logDetJ.squeeze(-1).squeeze(-1)
+
+        if self.residual == 'normal':
+            y = y + x
+            self._cached_logDetJ = F.softplus(self._cached_logDetJ)
+        elif self.residual == 'gated':
+            y = self.gate.sigmoid() * x + (1. - self.gate.sigmoid()) * y
+            term1 = torch.log(self.gate.sigmoid() + eps)
+            log1p_gate = torch.log1p(eps - self.gate.sigmoid())
+            log_gate = torch.log(self.gate.sigmoid() + eps)
+            term2 = F.softplus(log1p_gate - log_gate + self._cached_logDetJ)
+            self._cached_logDetJ = term1 + term2
+
+        return y
+
+    def _inverse(self, y):
+        """
+        :param y: the output of the bijection
+        :type y: torch.Tensor
+
+        Inverts y => x. As noted above, this implementation is incapable of inverting arbitrary values
+        `y`; rather it assumes `y` is the result of a previously computed application of the bijector
+        to some `x` (which was cached on the forward call)
+        """
+
+        raise KeyError("BlockNAFFlow expected to find key in intermediates cache but didn't")
+
+    def log_abs_det_jacobian(self, x, y):
```

Comment:
This seems a bit misleading since this always returns the last cached log det term regardless of x and y. Consider asserting that x and y are the same as in the cache, and throwing an exception otherwise.

Reply:
I agree, although this function should only be called from `torch.distributions.transform.TransformedDistribution`. I'll add the check!

Reply:
There are other parts of the code that call `log_abs_det_jacobian` directly, e.g. in HMC, though we probably wouldn't use this transform directly.

Reply:
I've made this change!